acpi/src/lib.rs

```diff
@@ -115,170 +138,150 @@ where
-        dsdt: Option<AmlTable>,
-        ssdts: Vec<AmlTable>,
-    ) -> AcpiTables<H> {
-        AcpiTables { revision, sdts, dsdt, ssdts, handler }
-    }
-
-    fn process_sdt(&mut self, physical_address: usize) -> Result<(), AcpiError> {
-        let header = sdt::peek_at_sdt_header(&self.handler, physical_address);
-        trace!("Found ACPI table with signature {:?} and length {:?}", header.signature, { header.length });
-
-        match header.signature {
-            Signature::FADT => {
-                use crate::fadt::Fadt;
-
-                /*
-                 * For whatever reason, they chose to put the DSDT inside the FADT, instead of just listing it
-                 * as another SDT. We extract it here to provide a nicer public API.
-                 */
-                let fadt_mapping =
-                    unsafe { self.handler.map_physical_region::<Fadt>(physical_address, mem::size_of::<Fadt>()) };
-                fadt_mapping.validate()?;
-
-                let dsdt_address = fadt_mapping.dsdt_address()?;
-                let dsdt_header = sdt::peek_at_sdt_header(&self.handler, dsdt_address);
-                self.dsdt = Some(AmlTable::new(dsdt_address, dsdt_header.length));
-
-                /*
-                 * We've already validated the FADT to get the DSDT out, so it doesn't need to be done again.
-                 */
-                self.sdts
-                    .insert(Signature::FADT, Sdt { physical_address, length: header.length, validated: true });
-            }
-            Signature::SSDT => {
-                self.ssdts.push(AmlTable::new(physical_address, header.length));
+    /// Finds and returns the DSDT AML table, if it exists.
+    pub fn dsdt(&self) -> AcpiResult<AmlTable> {
+        self.find_table::<fadt::Fadt>().and_then(|fadt| {
+            struct Dsdt;
+            // ### Safety: Implementation properly represents a valid DSDT.
+            unsafe impl AcpiTable for Dsdt {
+                const SIGNATURE: Signature = Signature::DSDT;
+
+                fn header(&self) -> &sdt::SdtHeader {
+                    // ### Safety: DSDT will always be valid for an SdtHeader at its `self` pointer.
+                    unsafe { &*(self as *const Self as *const sdt::SdtHeader) }
+                }
             }
-            signature => {
-                self.sdts.insert(signature, Sdt { physical_address, length: header.length, validated: false });
-            }
-        }
 
-        Ok(())
+            let dsdt_address = fadt.dsdt_address()?;
+            let dsdt = unsafe { read_table::<H, Dsdt>(self.handler.clone(), dsdt_address)? };
+
+            Ok(AmlTable::new(dsdt_address, dsdt.header().length))
+        })
     }
 
-    /// Create a mapping to a SDT, given its signature. This validates the SDT if it has not already been
-    /// validated.
-    ///
-    /// ### Safety
-    /// The table's memory is naively interpreted as a `T`, and so you must be careful in providing a type that
```

Comment:
I think it's a bit of a shame to lose all of this explanation, especially the bit about using `ExtendedField` to abstract fields that are only present on certain versions of ACPI or later. Could we move some of this to the `AcpiTable` documentation?

Reply:
I agree. I've modified the documentation of `AcpiTable` to read as follows:
```rust
/// All types representing ACPI tables should implement this trait.
///
/// ### Safety
///
/// The table's memory is naively interpreted, so you must be careful in providing a type that
/// correctly represents the table's structure. Regardless of the provided type's size, the region mapped will
/// be the size specified in the SDT's header. Providing a table impl that is larger than this, *may* lead to
/// page-faults, aliasing references, or derefencing uninitialized memory (the latter two being UB).
/// This isn't forbidden, however, because some tables rely on the impl being larger than a provided SDT in some
/// versions of ACPI (the [`ExtendedField`](crate::sdt::ExtendedField) type will be useful if you need to do
/// this. See our [`Fadt`](crate::fadt::Fadt) type for an example of this).
pub unsafe trait AcpiTable {
```
Is that acceptable?

Reply:
Yep, great - thanks!